compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -197,6 +236,138 @@ private void performCodeGen(boolean shrink) {
         codeGenCompleted = true;
     }
 
+    private void registerUnusedBIRNodes() {
+        List<ModuleContext> topologicallySortedModuleList = pkgResolution.topologicallySortedModuleList();
+        // Reversed the for loop because used BIRNode analysis should start from the root module.
+        // Root module is usually found last in the topologicallySortedModuleList.
+        for (int i = topologicallySortedModuleList.size() - 1; i >= 0; i--) {
+            ModuleContext moduleContext = topologicallySortedModuleList.get(i);
+
+            // Default module is analyzed first to find its immediate dependencies.
+            // Its immediate dependent modules are marked as "used" and they are optimized after that.
+            // This process happens till all "used" modules are exhausted.
+            if (shouldOptimize(moduleContext) && (isRootModule(moduleContext) || moduleContext.isUsed())) {
+                this.usedBIRNodeAnalyzer.analyze(moduleContext.bLangPackage());
+                updateNativeDependencyMap(moduleContext);
+            }
+        }
+
+        CodeGenOptimizationReportEmitter.emitBirOptimizationDuration();
+
+        if (this.packageContext.project().buildOptions().verbose()) {
+            CodeGenOptimizationReportEmitter.emitCodegenOptimizationReport(
+                    this.usedBIRNodeAnalyzer.pkgWiseInvocationData, getOptimizationReportPath(),
```

Comment:
You can pass the target path directly using `project.targetDir`

Reply:
We can apply the same changes here as in https://github.com/ballerina-platform/ballerina-lang/pull/41449#discussion_r1639715366 after finalizing it.

Reply:
Fixed in c2811f5f6a44844c14f04bba54927f8bdfbd5838.